File: src/com/trollworks/gcs/prereq/SkillPrereq.java
/*
 * Copyright (c) 1998-2015 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * version 2.0. If a copy of the MPL was not distributed with this file, You
 * can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as defined
 * by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.prereq;

import com.trollworks.toolkit.annotation.Localize;
import com.trollworks.toolkit.utility.Localization;


import com.trollworks.gcs.character.GURPSCharacter;
import com.trollworks.gcs.criteria.IntegerCriteria;
import com.trollworks.gcs.criteria.StringCompareType;
import com.trollworks.gcs.criteria.StringCriteria;
import com.trollworks.gcs.skill.Skill;
import com.trollworks.gcs.widgets.outline.ListRow;
import com.trollworks.toolkit.io.xml.XMLReader;
import com.trollworks.toolkit.io.xml.XMLWriter;

import java.io.IOException;
import java.text.MessageFormat;
import java.util.HashMap;
import java.util.HashSet;

/** A Skill prerequisite. */
public class SkillPrereq extends NameLevelPrereq {
	@Localize("{0}{1} a skill whose name {2}")
	@Localize(locale = "de", value = "{0}{1} eine Fertigkeit, deren Name {2}")
	@Localize(locale = "ru", value = "{0}{1} умение с названием {2}")
	private static String SKILL_NAME_PART;
	@Localize(", specialization {0},")
	@Localize(locale = "de", value = ", Spezialisierung {0},")
	@Localize(locale = "ru", value = ", специализация {0},")
	private static String SPECIALIZATION_PART;
	@Localize(" and level {0}")
	@Localize(locale = "de", value = " und Fertigkeitswert {0}")
	@Localize(locale = "ru", value = " и уровень {0}\n ")
	private static String LEVEL_PART;
	@Localize(" level {0} and tech level matches\n")
	@Localize(locale = "de", value = " Fertigkeitswert {0} und Techlevel stimmt überein")
	@Localize(locale = "ru", value = " уровень {0} и ТУ совпадают\n")
	private static String LEVEL_AND_TL_PART;

	static {
		Localization.initialize();
	}

	/** The XML tag for this class. */
	public static final String	TAG_ROOT			= "skill_prereq";	//$NON-NLS-1$
	private static final String	TAG_SPECIALIZATION	= "specialization"; //$NON-NLS-1$
	private StringCriteria		mSpecializationCriteria;

	/**
	 * Creates a new prerequisite.
	 * 
	 * @param parent The owning prerequisite list, if any.
	 */
	public SkillPrereq(PrereqList parent) {
		super(TAG_ROOT, parent);
		mSpecializationCriteria = new StringCriteria(StringCompareType.IS_ANYTHING, ""); //$NON-NLS-1$
	}

	/**
	 * Loads a prerequisite.
	 * 
	 * @param parent The owning prerequisite list, if any.
	 * @param reader The XML reader to load from.
	 */
	public SkillPrereq(PrereqList parent, XMLReader reader) throws IOException {
		super(parent, reader);
	}

	private SkillPrereq(PrereqList parent, SkillPrereq prereq) {
		super(parent, prereq);
		mSpecializationCriteria = new StringCriteria(prereq.mSpecializationCriteria);
	}

	@Override
	protected void initializeForLoad() {
		mSpecializationCriteria = new StringCriteria(StringCompareType.IS_ANYTHING, ""); //$NON-NLS-1$
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == this) {
			return true;
		}
		if (obj instanceof SkillPrereq && super.equals(obj)) {
			return mSpecializationCriteria.equals(((SkillPrereq) obj).mSpecializationCriteria);
		}
		return false;
	}

	@Override
	protected void loadSelf(XMLReader reader) throws IOException {
		if (TAG_SPECIALIZATION.equals(reader.getName())) {
			mSpecializationCriteria.load(reader);
		} else {
			super.loadSelf(reader);
		}
	}

	@Override
	protected void saveSelf(XMLWriter out) {
		mSpecializationCriteria.save(out, TAG_SPECIALIZATION);
	}

	@Override
	public String getXMLTag() {
		return TAG_ROOT;
	}

	@Override
	public Prereq clone(PrereqList parent) {
		return new SkillPrereq(parent, this);
	}

	@Override
	public boolean satisfied(GURPSCharacter character, ListRow exclude, StringBuilder builder, String prefix) {
		boolean satisfied = false;
		String techLevel = null;
		StringCriteria nameCriteria = getNameCriteria();
		IntegerCriteria levelCriteria = getLevelCriteria();

		if (exclude instanceof Skill) {
			techLevel = ((Skill) exclude).getTechLevel();
		}

		for (Skill skill : character.getSkillsIterator()) {
			if (exclude != skill && nameCriteria.matches(skill.getName()) && mSpecializationCriteria.matches(skill.getSpecialization())) {
				satisfied = levelCriteria.matches(skill.getLevel());
				if (satisfied && techLevel != null) {
					String otherTL = skill.getTechLevel();
					satisfied = otherTL == null || techLevel.equals(otherTL);
				}
				break;
			}
		}
		if (!has()) {
			satisfied = !satisfied;
		}
		if (!satisfied && builder != null) {
			builder.append(MessageFormat.format(SKILL_NAME_PART, prefix, has() ? HAS : DOES_NOT_HAVE, nameCriteria.toString()));
			boolean notAnySpecialization = mSpecializationCriteria.getType() != StringCompareType.IS_ANYTHING;

			if (notAnySpecialization) {
				builder.append(MessageFormat.format(SPECIALIZATION_PART, mSpecializationCriteria.toString()));
			}
			if (techLevel == null) {
				builder.append(MessageFormat.format(LEVEL_PART, levelCriteria.toString()));
			} else {
				if (notAnySpecialization) {
					builder.append(","); //$NON-NLS-1$
				}
				builder.append(MessageFormat.format(LEVEL_AND_TL_PART, levelCriteria.toString()));
			}
		}
		return satisfied;
	}

	@Override
	public void fillWithNameableKeys(HashSet<String> set) {
		super.fillWithNameableKeys(set);
		ListRow.extractNameables(set, mSpecializationCriteria.getQualifier());
	}

	@Override
	public void applyNameableKeys(HashMap<String, String> map) {
		super.applyNameableKeys(map);
		mSpecializationCriteria.setQualifier(ListRow.nameNameables(map, mSpecializationCriteria.getQualifier()));
	}

	/** @return The specialization comparison object. */
	public StringCriteria getSpecializationCriteria() {
		return mSpecializationCriteria;
	}
}

File: model/gurps/skill_prereq.go
// Copyright (c) 1998-2025 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package gurps

import (
	"hash"

	"github.com/richardwilkes/gcs/v5/model/criteria"
	"github.com/richardwilkes/gcs/v5/model/gurps/enums/prereq"
	"github.com/richardwilkes/gcs/v5/model/nameable"
	"github.com/richardwilkes/toolbox/i18n"
	"github.com/richardwilkes/toolbox/xio"
	"github.com/richardwilkes/toolbox/xmath/hashhelper"
)

var _ Prereq = &SkillPrereq{}

// SkillPrereq holds a prerequisite for a skill.
type SkillPrereq struct {
	Parent                 *PrereqList     `json:"-"`
	Type                   prereq.Type     `json:"type"`
	Has                    bool            `json:"has"`
	NameCriteria           criteria.Text   `json:"name,omitempty"`
	LevelCriteria          criteria.Number `json:"level,omitempty"`
	SpecializationCriteria criteria.Text   `json:"specialization,omitempty"`
}

// NewSkillPrereq creates a new SkillPrereq.
func NewSkillPrereq() *SkillPrereq {
	var p SkillPrereq
	p.Type = prereq.Skill
	p.NameCriteria.Compare = criteria.IsText
	p.LevelCriteria.Compare = criteria.AtLeastNumber
	p.SpecializationCriteria.Compare = criteria.AnyText
	p.Has = true
	return &p
}

// PrereqType implements Prereq.
func (p *SkillPrereq) PrereqType() prereq.Type {
	return p.Type
}

// ParentList implements Prereq.
func (p *SkillPrereq) ParentList() *PrereqList {
	return p.Parent
}

// Clone implements Prereq.
func (p *SkillPrereq) Clone(parent *PrereqList) Prereq {
	clone := *p
	clone.Parent = parent
	return &clone
}

// FillWithNameableKeys implements Prereq.
func (p *SkillPrereq) FillWithNameableKeys(m, existing map[string]string) {
	nameable.Extract(p.NameCriteria.Qualifier, m, existing)
	nameable.Extract(p.SpecializationCriteria.Qualifier, m, existing)
}

// Satisfied implements Prereq.
func (p *SkillPrereq) Satisfied(entity *Entity, exclude any, tooltip *xio.ByteBuffer, prefix string, _ *bool) bool {
	var replacements map[string]string
	if na, ok := exclude.(nameable.Accesser); ok {
		replacements = na.NameableReplacements()
	}
	satisfied := false
	var techLevel *string
	if sk, ok := exclude.(*Skill); ok {
		techLevel = sk.TechLevel
	}
	Traverse(func(sk *Skill) bool {
		if exclude == sk || !p.NameCriteria.Matches(replacements, sk.NameWithReplacements()) ||
			!p.SpecializationCriteria.Matches(replacements, sk.SpecializationWithReplacements()) {
			return false
		}
		satisfied = p.LevelCriteria.Matches(sk.LevelData.Level)
		if satisfied && techLevel != nil {
			satisfied = sk.TechLevel == nil || *techLevel == *sk.TechLevel
		}
		return satisfied
	}, false, true, entity.Skills...)
	if !p.Has {
		satisfied = !satisfied
	}
	if !satisfied && tooltip != nil {
		tooltip.WriteString(prefix)
		tooltip.WriteString(HasText(p.Has))
		tooltip.WriteString(i18n.Text(" a skill whose name "))
		tooltip.WriteString(p.NameCriteria.String(replacements))
		if p.SpecializationCriteria.Compare != criteria.AnyText {
			tooltip.WriteString(i18n.Text(", specialization "))
			tooltip.WriteString(p.SpecializationCriteria.String(replacements))
			tooltip.WriteByte(',')
		}
		if techLevel == nil {
			tooltip.WriteString(i18n.Text(" and level "))
			tooltip.WriteString(p.LevelCriteria.String())
		} else {
			if p.SpecializationCriteria.Compare != criteria.AnyText {
				tooltip.WriteByte(',')
			}
			tooltip.WriteString(i18n.Text(" level "))
			tooltip.WriteString(p.LevelCriteria.String())
			tooltip.WriteString(i18n.Text(" and tech level matches"))
		}
	}
	return satisfied
}

// Hash writes this object's contents into the hasher.
func (p *SkillPrereq) Hash(h hash.Hash) {
	if p == nil {
		hashhelper.Num8(h, uint8(255))
		return
	}
	hashhelper.Num8(h, p.Type)
	hashhelper.Bool(h, p.Has)
	p.NameCriteria.Hash(h)
	p.LevelCriteria.Hash(h)
	p.SpecializationCriteria.Hash(h)
}
